### PR TITLE
feat(anta): Added test case to verify SNMP Trap PDUs counters

### LINKED ISSUE #820: Add test case to verify SNMP counters
Trap PDUs counter should be a non-zero value

sample output:

leaf1-dc1#show snmp | json
{
    "counters": {
        "serviceEnabled": true,
        "inPkts": 0,
        "inVersionErrs": 0,
        "inBadCommunityNames": 0,
        "inBadCommunityUses": 0,
        "inParseErrs": 0,
        "inRequestVars": 0,
        "inSetVars": 0,
        "inGetPdus": 0,
        "inGetNextPdus": 0,
        "inSetPdus": 0,
        "outPkts": 26,
        "outTooBigErrs": 0,
        "outNoSuchNameErrs": 0,
        "outBadValueErrs": 0,
        "outGeneralErrs": 0,
        "outGetResponsePdus": 0,
        "outTrapPdus": 26,
        "outTrapDrops": 0,
        "groupLength": 0,
        "numberOfGroups": 0,
        "numberOfViews": 0
    },

### PR DESCRIPTION
**Pull request by vitthalmagadum**
# Description

Verifies the number of SNMP PDU(s) (Protocol Data Units) processed.

    By default, all SNMP PDU counters will be checked for any non-zero values.
    An optional list of specific SNMP PDU(s) can be provided for granular testing.

    Expected Results
    ----------------
    * Success: The test will pass if the SNMP PDU counter(s) are non-zero/greater than zero.
    * Failure: The test will fail if the SNMP PDU counter(s) are zero/None/Not Found.

Fixes #820 

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have run pre-commit for code linting and typing (`pre-commit run`)
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes (`tox -e testenv`)

### CODE REVIEW

File: anta/tests/snmp.py [hunk 1]
```diff
@@ -237,3 +240,60 @@ def test(self) -> None:
             self.result.is_failure(f"Expected `{self.inputs.contact}` as the contact, but found `{contact}` instead.")
         else:
             self.result.is_success()
+
+
+class VerifySNMPPDUs(AntaTest):
```
[the remark this answers is not in the document]
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Updated.

File: anta/tests/snmp.py [hunk 2]
```diff
@@ -237,3 +240,60 @@ def test(self) -> None:
...
+
+    @AntaTest.anta_test
+    def test(self) -> None:
+        """Main test function for VerifySNMPPDUs."""
+        snmp_pdus = self.inputs.pdus
+        command_output = self.instance_commands[0].json_output
+
+        # Verify SNMP PDU counters.
+        if not (pdu_counters := get_value(command_output, "counters")):
+            self.result.is_failure("SNMP counter details not found.")
```
[the remark this answers is not in the document]
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Updated the failure message

File: anta/tests/snmp.py [hunk 3]
```diff
@@ -237,3 +240,60 @@ def test(self) -> None:
...
+        if not snmp_pdus:
+            snmp_pdus = SNMP_PDUS
+
+        failures = [pdu for pdu in snmp_pdus if not pdu_counters.get(pdu)]
+
+        # Check if any failures
+        if not failures:
+            self.result.is_success()
+        else:
+            self.result.is_failure(f"The following SNMP PDU(s) are not found or have zero PDU counter:\n{', '.join(failures)}")
```
[the remark this answers is not in the document]
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Updated the failure message. Thanks!!

File: tests/units/anta_tests/test_snmp.py [hunk 4]
```diff
@@ -152,4 +152,66 @@
...
+                    "inGetPdus": 0,
+                    "inGetNextPdus": 2,
+                    "inSetPdus": 0,
+                    "outGetResponsePdus": 3,
+                    "outTrapPdus": 9,
+                },
+            }
+        ],
+        "inputs": {},
+        "expected": {"result": "failure", "messages": ["The following SNMP PDU(s) are not found or have zero PDU counter:\n{'inGetPdus': 0, 'inSetPdus': 0}"]},
```
Comment (MaheshGSLAB) [inline tests/units/anta_tests/test_snmp.py:215]: please add the case where pdu is not found

Comment (vitthalmagadum) [reply tests/units/anta_tests/test_snmp.py:215]: Added check. Thank you!!

File: anta/tests/snmp.py [hunk 5]
```diff
@@ -237,3 +242,73 @@ def test(self) -> None:
             self.result.is_failure(f"Expected `{self.inputs.contact}` as the contact, but found `{contact}` instead.")
         else:
             self.result.is_success()
+
+
+class VerifySnmpPDUs(AntaTest):
+    """Verifies the number of SNMP PDU(s) (Protocol Data Units) processed.
```
Comment (carl-baillargeon) [inline anta/tests/snmp.py:248]:
```suggestion
    """Verifies the SNMP PDU (Protocol Data Unit) counters.
```

File: anta/tests/snmp.py [hunk 6]
```diff
@@ -237,3 +242,73 @@ def test(self) -> None:
             self.result.is_failure(f"Expected `{self.inputs.contact}` as the contact, but found `{contact}` instead.")
         else:
             self.result.is_success()
+
+
+class VerifySnmpPDUs(AntaTest):
```
Comment (carl-baillargeon) [inline anta/tests/snmp.py:247]:
```suggestion
class VerifySnmpPDUCounters(AntaTest):
```

File: anta/tests/snmp.py [hunk 7]
```diff
@@ -237,3 +242,73 @@ def test(self) -> None:
...
+    anta.tests.snmp:
+      - VerifySnmpPDUs:
+          pdus:
+            - outTrapPdus
+            - inGetNextPdus
+    ```
+    """
+
+    name = "VerifySnmpPDUs"
+    description = "Verifies the number of SNMP PDU(s) (Protocol Data Units) processed."
```
Comment (carl-baillargeon) [inline anta/tests/snmp.py:270]:
```suggestion
    description = "Verifies the SNMP PDU (Protocol Data Unit) counters."
```

File: anta/tests/snmp.py (hunk 6 above)
Comment (carl-baillargeon) [reply anta/tests/snmp.py:247]: Also change the docstrings, etc.

File: anta/tests/snmp.py [hunk 8]
```diff
@@ -237,3 +242,73 @@ def test(self) -> None:
...
+    categories: ClassVar[list[str]] = ["snmp"]
+    commands: ClassVar[list[AntaCommand | AntaTemplate]] = [AntaCommand(command="show snmp", revision=1)]
+
+    class Input(AntaTest.Input):
+        """Input model for the VerifySnmpPDUs test."""
+
+        pdus: list[str] | None = None
```
Comment (carl-baillargeon) [inline anta/tests/snmp.py:277]:
```suggestion
        pdus: list[SnmpPdu] | None = None
```

File: anta/tests/snmp.py [hunk 9]
```diff
@@ -237,3 +242,73 @@ def test(self) -> None:
...
+        """Optional list of SNMP PDU counters to be verified. If not provided, test will verifies all the PDUs."""
```
Comment (carl-baillargeon) [inline anta/tests/snmp.py:278]:
```suggestion
        """Optional list of SNMP PDU counters to be verified. If not provided, test will verifies all PDU counters."""
```

File: anta/tests/snmp.py [hunk 10]
```diff
@@ -237,3 +242,73 @@ def test(self) -> None:
...
+            """Validate the inputs provided to the VerifySnmpPDUs test.
+
+            The valid SNMP PDU counter must be provided.
+            """
+            if self.pdus:
+                for pdu in self.pdus:
+                    if pdu not in SNMP_PDUS:
+                        msg = f"Invalid PDU counter {pdu}. Must be one of {SNMP_PDUS}."
+                        raise ValueError(msg)
+            return self
```
Comment (carl-baillargeon) [inline anta/tests/snmp.py:291]:
```suggestion
```

Not needed with a custom type :)

File: anta/tests/snmp.py [hunk 11]
```diff
@@ -237,3 +242,73 @@ def test(self) -> None:
...
+
+    @AntaTest.anta_test
+    def test(self) -> None:
+        """Main test function for VerifySnmpPDUs."""
+        snmp_pdus = self.inputs.pdus
+        command_output = self.instance_commands[0].json_output
+
+        # Verify SNMP PDU counters.
+        if not (pdu_counters := get_value(command_output, "counters")):
+            self.result.is_failure("SNMP counter details are not found.")
```
Comment (carl-baillargeon) [inline anta/tests/snmp.py:301]:
```suggestion
            self.result.is_failure("SNMP counters not found.")
```

File: anta/tests/snmp.py [hunk 12]
```diff
@@ -237,3 +240,60 @@ def test(self) -> None:
...
+        command_output = self.instance_commands[0].json_output
+
+        # Verify SNMP PDU counters.
+        if not (pdu_counters := get_value(command_output, "counters")):
+            self.result.is_failure("SNMP counter details not found.")
+            return
+
+        # In case SNMP PDUs not provided, It will check all the update error counters.
+        if not snmp_pdus:
+            snmp_pdus = SNMP_PDUS
```
[the remark this answers is not in the document]
Comment (carl-baillargeon) [reply anta/tests/snmp.py:306]: @vitthalmagadum Here you can use `typing.get_args` on `SnmpPdu` custom type instead.

File: anta/tests/snmp.py [hunk 13]
```diff
@@ -237,3 +242,73 @@ def test(self) -> None:
...
+
+        failures = {pdu: value for pdu in snmp_pdus if (value := pdu_counters.get(pdu, "Not Found")) == "Not Found" or value == 0}
+
+        # Check if any failures
+        if not failures:
+            self.result.is_success()
+        else:
+            self.result.is_failure(f"The following SNMP PDU(s) are not found or have zero PDU counter:\n{failures}")
```
Comment (carl-baillargeon) [inline anta/tests/snmp.py:314]:
```suggestion
            self.result.is_failure(f"The following SNMP PDU counters are not found or have zero PDU counters:\n{failures}")
```

File: anta/tests/snmp.py [hunk 14]
```diff
@@ -9,13 +9,18 @@
...
+
 from anta.custom_types import PositiveInteger
 from anta.models import AntaCommand, AntaTest
 from anta.tools import get_value
 
 if TYPE_CHECKING:
     from anta.models import AntaTemplate
 
+# Define the valid PDUs as a list
+SNMP_PDUS: list[str] = ["inGetPdus", "inGetNextPdus", "inSetPdus", "outGetResponsePdus", "outTrapPdus"]
```
[the remark this answers is not in the document]
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Added into custom types.
Thanks!!

File: anta/tests/snmp.py (hunk 5 above)
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Updated.

File: anta/tests/snmp.py (hunk 6 above)
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Updated test name and docstring as well.

File: anta/tests/snmp.py (hunk 7 above)
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Updated.

File: anta/tests/snmp.py (hunk 8 above)
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Updated.

File: anta/tests/snmp.py (hunk 9 above)
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Updated.

File: anta/tests/snmp.py (hunk 10 above)
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Removed validator.

File: anta/tests/snmp.py (hunk 11 above)
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Updated.

File: anta/tests/snmp.py (hunk 13 above)
Comment (vitthalmagadum) [reply anta/tests/snmp.py]: Updated.

**Review by MaheshGSLAB — APPROVED** [review]

**Review by carl-baillargeon — APPROVED** [review]
LGTM

**Pull request merged**